src/wam2layers/preprocessing/shared.py

```diff
@@ -1,218 +1,164 @@
-    """Infer the lat and lon bounds and add to the dataset (in-place)."""
-    lats = ds["latitude"].to_numpy()
-    lons = ds["longitude"].to_numpy()
-    res_lat = np.median((np.diff(lats)))  # infer resolution
-    res_lon = np.median((np.diff(lons)))
-
-    ds["latitude_bnds"] = (
-        ("latitude", "bnds"),
-        np.array((lats - res_lat / 2, lats + res_lat / 2)).T,
-    )
-    ds["longitude_bnds"] = (
-        ("longitude", "bnds"),
-        np.array((lons - res_lon / 2, lons + res_lon / 2)).T,
-    )
+    config = Config.from_yaml(config_file)
+
+    for datetime in get_input_dates(config):
+        is_new_day = datetime == datetime.floor("1d")
+        logger.info(datetime)
+
+        # 4d fields
+        levels = config.levels
+
+        input_data = xr.Dataset()
+        input_data["q"] = load_data("q", datetime, config)  # in kg kg-1
+        input_data["u"] = load_data("u", datetime, config)  # in m/s
+        input_data["v"] = load_data("v", datetime, config)  # in m/s
+
+        surface_data = xr.Dataset()
+        surface_data["ps"] = load_data("sp", datetime, config)  # in Pa
+
+        if config.level_type == "model_levels":
+            dp = get_dp_modellevels(surface_data["ps"], levels)
+
+        if config.level_type == "pressure_levels":
+            d2m = load_data("d2m", datetime, config)  # Dew point in K
+            surface_data["qs"] = calculate_humidity(d2m, surface_data["ps"])  # kg kg-1
+            surface_data["us"] = load_data("u10", datetime, config)  # in m/s
+            surface_data["vs"] = load_data("v10", datetime, config)  # in m/s
+
+            input_data, pb = extend_pressurelevels(input_data, surface_data)
+            input_data, dp = interp_dp_midpoints(input_data, surface_data["ps"])
+
+        # Calculate column water vapour
+        g = 9.80665  # gravitational accelleration [m/s2]
+        cwv = input_data["q"] * dp / g  # (kg/m2)
+
+        try:
+            # Calculate column water instead of column water vapour
+            tcw = load_data("tcw", datetime, config)  # kg/m2
+            correction = tcw / cwv.sum(dim="level")
+            cw = correction * cwv  # column water (kg/m2)
+            if is_new_day:
+                logger.info(
+                    "Total column water correction:\n"
+                    "    ratio total column water / computed integrated water vapour\n"
+                    f"    mean over grid for this timestep {correction.mean().item():.4f}"
+                )
+
+        except FileNotFoundError:
+            # Fluxes will be calculated based on the column water vapour
+            cw = cwv
+            log_once(
+                logger, "Total column water not available; using water vapour only"
+            )
+
+        # Integrate fluxes and states to upper and lower layer
+        if config.level_type == "model_levels":
+            # TODO: Check if this is a reasonable choice for boundary
+            boundary = 111
```

Comment:
This is quite specific for ERA5 only!

Reply:
Yeah, I did note that down in my comment. We can discuss this tomorrow

Reply:
Made this generic, and an optional model config